v1/nix/modules/drv-parts/nodejs-devshell/default.nix

```diff
@@ -0,0 +1,57 @@
+{
+  config,
+  lib,
+  dream2nix,
+  packageSets,
+  ...
+}: let
+  l = lib // builtins;
+
+  cfg = config.nodejs-devshell;
+
+  nodeModulesDrv = dream2nix.lib.evalModules {
+    inherit packageSets;
+    modules = [
+      dream2nix.modules.drv-parts.nodejs-node-modules
+      {inherit (config) nodejs-package-lock name version;}
+      {mkDerivation.src = l.mkForce null;}
+    ];
+  };
+
+  nodeModulesDir = "${nodeModulesDrv}/lib/node_modules/${config.name}/node_modules";
+  binDir = "${nodeModulesDrv}/lib/node_modules/.bin";
+in {
+  imports = [
+    dream2nix.modules.drv-parts.mkDerivation
+    dream2nix.modules.drv-parts.nodejs-package-lock
+  ];
+
+  # rsync the node_modules folder
+  # - tracks node-modules store path via .dream2nix/.node_modules_id
+  # - omits copying if store path did not change
+  # - if store path changes, only replaces updated files
+  # - rsync can be restarted from any point, if failed or aborted mid execution.
+  # Options:
+  # -a            -> all files recursive, preserve symlinks, etc.
+  # --delete      -> removes deleted files
+  # --chmod=+ug+w -> make folder writeable by user+group
+  mkDerivation.shellHook = ''
+    ID=${nodeModulesDir}
+    currID=$(cat .dream2nix/.node_modules_id 2> /dev/null)
+
+    mkdir -p .dream2nix
+    if [[ "$ID" != "$currID" || ! -d "node_modules"  ]];
+    then
+      ${config.deps.rsync}/bin/rsync -a --chmod=ug+w  --delete ${nodeModulesDir}/ ./node_modules/
+      mkdir -p ./node_modules/.bin
+      for executablePath in ${binDir}/*; do
```

Comment:
I agree and don't understand why it's necessary here, if it is?

> checked if ln succeeds when a link with the same name is already present? (I think it fails and then you have to do many conditional stuff here)

That one alone could be solved with `ln -sf` :)

Reply:
Providing some background, the granular builder currently installs the top-level dependency inside a top-level node_modules. In a way you could say that the builder treats the package as a dependency of itself. Therefore it creates the `.bin` entry one layer up (to high for our use case here).

If this was a smart choice or not can be debated, but changing this requires larger refactoring.

Also I think that it's hard to find a `correct` generic way of building the .bin directory, because it's content depends on multiple variables like: where the package is located in the tree, how the tree is built. This information is not yet known when building an individual dependency, therefore harder to pre-build/cache it.

Therefore, depending on which settings we choose for the builder, like `installMethod` fore example, the logic for generating .bin has to change.

Via 67cf11ad1befd30439b4e358a6a043040307733b I now improved the `nodejs-node-modules` module to already build the `.bin` in the way we want to have it for a dev-shell, and therefore removed this logic from the devshell module.

Reply:
... this allows relocating the node_modules via a single rsync